### PR TITLE
Symlink hermetic cuda headers to permit clang cuda version detection

### LINKED ISSUE #16877: Clang cannot detect hermetic cuda version
When using the hermetic CUDA toolchain with Clang, the CUDA version is not actually detected by Clang and defaults to the latest known toolkit version. Clang searches for `external/cuda_nvcc/include/cuda.h`, and when it is not found, the version detection logic switches to `CudaVersion::NEW`.

For instance, in clang 18.1.8, CudaVersion::NEW is 12.3. When we set `HERMETIC_CUDA_VERSION="12.2.0"`, this mismatch and failed detection causes clang to use isa 8.3 which is incompatible with cuda 12.2.0. 

The easiest fix is probably to copy `external/cuda_cudart/include/cuda.h` into `external/cuda_nvcc/include/cuda.h`.

### PR DESCRIPTION
**Pull request by tchatow**
Fixes #16877

### CODE REVIEW

**Comment by beckerhe** [comment]
Looks good to me, but I would like @ybaturina to have a look.

File: third_party/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl [hunk 1]
```diff
@@ -218,6 +218,10 @@ def _create_libcuda_symlinks(
         repository_ctx.symlink(nvidia_driver_path, "lib/libcuda.so.1")
         repository_ctx.symlink("lib/libcuda.so.1", "lib/libcuda.so")
 
+def _create_cuda_header_symlinks(repository_ctx):
+    if repository_ctx.name == "cuda_nvcc":
+        repository_ctx.symlink("../cuda_cudart/include/cuda.h", "include/cuda.h")
```
Comment (ybaturina) [inline third_party/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:223]: Would you explain the algorithm here please, e.g. how does it work?
When does Clang use this symlink?

Comment (tchatow) [reply third_party/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:223]: Clang uses `cuda.h` in a location relative to nvcc to determine the cuda version. If it is missing, Clang assumes the latest known Cuda version which is usually incorrect. The hermetic toolchain breaks this by downloading nvcc and cudart to separate directories.

See: https://github.com/llvm/llvm-project/blob/b592917eec407a01ef07334af5aafe6c8922c80a/clang/lib/Driver/ToolChains/Cuda.cpp#L212

Comment (ybaturina) [reply third_party/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:223]: Thank you. Are you planning to use this change for CLANG builds (not CLANG+NVCC)?

Comment (tchatow) [reply third_party/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:223]: Yes, this is for CLANG builds.

Comment (tchatow) [reply third_party/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:223]: I think this change merely exposes an already broken build configuration. In your case you should set `-Wno-unknown-cuda-version` to get the same behavior as today. Defaulting to `CudaVersion::NEW` completely ignores the hermetic nvcc version and causes isa generation incompatible with ptxas.

Comment (ybaturina) [reply third_party/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:223]: In this case may I ask you to change .bazelrc file and add the following to the `build:cuda_clang` config:

`build:cuda_clang --copt="-Wno-unknown-cuda-version"`

And then I'll be able to approve it.

**Comment by thomasjoerg** [comment]
ybaturina is OOO. 
@beckerhe do you have enough context to take over?

**Comment by beckerhe** [comment]
> ybaturina is OOO. @beckerhe do you have enough context to take over?

I believe Yulia's ask from the comment chain above still stands. Once that's resolved it can be merged.

**Comment by beckerhe** [comment]
Thank you. I think this is good to go but would like to trigger a new run of the CI. Unfortunately the PR needs to be synced to HEAD due to a breakage on main that has been fixed by now.

I don't have the permission to sync the PR. @thomasjoerg or @dimitar-asenov, can you sync the PR and approve a run of the CI?

**Review by beckerhe — APPROVED** [review]

**Comment by tchatow** [comment]
Rebased

**Comment by dimitar-asenov** [comment]
Thanks, Sorry for the delay on submitting this. I hope to have it in today.

**Review by dimitar-asenov — APPROVED** [review]

**Pull request closed without merging**